kratos/includes/model_part.h

```diff
@@ -1525,43 +1466,69 @@ class KRATOS_API(KRATOS_CORE) ModelPart final
                     KRATOS_ERROR << "Attempting to add a new geometry with Id: " << it_found->Id() << ". A different geometry with the same Id already exists." << std::endl;
-                } else {
-                    aux.push_back(*(it.base()));
                 }
             }
-        }
+        });
 
-        // Add to root model part
-        for(auto& p_geom : aux_root) {
-            p_root_model_part->AddGeometry(p_geom);
-        }
+        InsertEntityRange([](ModelPart* pModelPart) {
+            return &(pModelPart->Geometries());
+        }, GeometryBegin, GeometriesEnd);
+
+        KRATOS_CATCH("")
+    }
 
-        // Add to all of the leaves
-        ModelPart* p_current_part = this;
-        while(p_current_part->IsSubModelPart()) {
-            for(auto& p_geom : aux) {
-                p_current_part->AddGeometry(p_geom);
-            }
 
-            p_current_part = &(p_current_part->GetParentModelPart());
+    /// Inserts a list of geometries to a submodelpart provided their iterators
+    template<class TContainer, std::enable_if_t<IsRValueContainer<TContainer>::value, bool> = true>
+    void AddGeometries(TContainer&& rInputContainer)
+    {
+        KRATOS_TRY
+
+        // This check is required in cases where std::vector<IndexType> is passed as an r value, then it will
+        // not call the correct addition with the input args of const std::vector<IndexType>&, it will get in to this method.
+        // this check will redirect it to the std::vector<IndexType> method with the rContainer.
+        if constexpr(!std::is_same_v<BaseType<TContainer>, std::vector<IndexType>>) {
+            ModelPart* p_root_model_part = &this->GetRootModelPart();
+
+            block_for_each(rInputContainer.begin(), rInputContainer.end(), [p_root_model_part](const auto& prGeometry) {
+                const auto& r_geometry = ReferenceGetter<GeometriesMapType::value_type>::Get(prGeometry);
+                const auto& r_geometries = p_root_model_part->Geometries();
+                auto it_found = r_geometries.find(r_geometry.Id());
+                if (it_found != p_root_model_part->GeometriesEnd()) {
+                    if (GeometryType::HasSameGeometryType(r_geometry, *it_found)) { // Check the geometry type and connectivities
+                        for (IndexType i_pt = 0; i_pt < r_geometry.PointsNumber(); ++i_pt) {
+                            KRATOS_ERROR_IF((r_geometry)[i_pt].Id() != (*it_found)[i_pt].Id()) << "Attempting to add a new geometry with Id: " << r_geometry.Id() << ". A same type geometry with same Id but different connectivities already exists." << std::endl;
+                        }
+                    } else if(&(*it_found) != &r_geometry) { // Check if the pointee coincides
+                        KRATOS_ERROR << "Attempting to add a new geometry with Id: " << it_found->Id() << ". A different geometry with the same Id already exists." << std::endl;
+                    }
```

Comment:
@sunethwarna I know you didn't write this and it's the current behavior, but I have to mention that I really don't like that we have to support duplicate insertions of the same geometry.

Apart from code complexity issues, this really **does** have a pretty significant performance impact on insertion. I don't get why we're nitpicking over other minor issues when we validate every single geometry inserted into the model part.

Reply:
i would say this is a different discussion ... i really do not think we should chaneg any behaviour in this PR